### ISSUE TITLE
Expose core::intrinsics::volatile_copy_nonoverlapping_memory as core::ptr::volatile_copy_nonoverlapping

### ISSUE THREAD

**Issue by joshlf** [rust-lang/rust#58041]
I propose that we expose [`core::intrinsics::volatile_copy_nonoverlapping_memory`](https://doc.rust-lang.org/core/intrinsics/fn.volatile_copy_nonoverlapping_memory.html) as the stable `core::ptr::volatile_copy_nonoverlapping`. It might make sense to stabilize other intrinsics while we're at it; I only mention this one in particular because I have a use case for it.

Is a discussion here sufficient, or should I submit an RFC?

**Comment by scottmcm** [comment]
Small libs changes often don't need RFCs, so this is probably fine.

Can you elaborate on why nonoverlapping is useful to you in volatile?  Much of its advantage over regular copy is in optimizations, many of which are turned off for volatile...

**Comment by joshlf** [comment]
I have a library which is used to safely operate on memory which is shared with another, untrusted process. We have to assume that the foreign process may be arbitrarily modifying the memory while we're operating on it, making normal memory accesses unsound. Volatile allows us to ensure that the compiler doesn't make any unsound optimizations with respect to reads or writes from/to the memory. More details [here](https://fuchsia-review.googlesource.com/c/garnet/+/229382).

**Comment by Amanieu** [comment]
I'm not a fan of `volatile_copy_memory` because it uses volatile accesses for both loads and stores, which is usually not what you want. It might be better to just use `read_volatile`/`write_volatile` in a loop.

**Comment by joshlf** [comment]
Oh that's a very good point. However, doing volatile ops in a loop can be TERRIBLE for performance because the compiler isn't allowed to coalesce them. Do you know if there are variable-length equivalents of `read_volatile`/`write_volatile`?

**Comment by scottmcm** [comment]
@joshlf Sorry, I was unclear: I meant why wrapping the `nonoverlapping` version instead of wrapping https://doc.rust-lang.org/core/intrinsics/fn.volatile_copy_memory.html?  Do the `nonoverlapping`-specific optimizations matter with volatile?  Might someone need the memmove one later?

Also, https://github.com/rust-lang/rust/issues/58041#issuecomment-459800070 reminds me of a URLO conversation about sharing memory:

> That said, I cannot think of any reason to actually use volatile here, or even think of any case where adding volatile removes UB.

[https://users.rust-lang.org/t/how-unsafe-is-mmap/19635/12](https://users.rust-lang.org/t/how-unsafe-is-mmap/19635/12?u=scottmcm)

**Comment by joshlf** [comment]
>  I meant why wrapping the `nonoverlapping` version

All of our uses are nonoverlapping. I'm not sure whether it actually provides any optimization opportunities, but I'm sure it can't hurt :)

> That said, I cannot think of any reason to actually use volatile here, or even think of any case where adding volatile removes UB.

Our use case is different - we're sharing memory with a remote process which is assumed to be malicious, and so we must assume that the memory is changing out from under us. Thus, accesses have to either be volatile or atomic. If they're not, it's equivalent to a data race, which is UB.

**Comment by Lokathor** [comment]
If I understand your situation correctly, atomic would not help you here. You'd need volatile.

**Comment by joshlf** [comment]
@Lokathor 

> If I understand your situation correctly, atomic would not help you here. You'd need volatile.

I don't believe that's right; so long as you use atomic operations, the compiler guarantees that concurrent modifications by a different thread will not cause UB. I can't imagine how that other thread being in a different process would change anything.

**Comment by Lokathor** [comment]
atomics can't be re-ordered, but as i understand it they can be elided according to [LLVM rules](https://llvm.org/docs/Atomics.html#id4):

> A couple examples: if a SequentiallyConsistent store is immediately followed by another SequentiallyConsistent store to the same address, the first store can be erased. This transformation is not allowed for a pair of volatile stores.

**Comment by joshlf** [comment]
Atomics being reordered is fine; this is about a security rather than a correctness property (if the remote process is misbehaving, correctness is already out the window). In fact, even reordering would be fine for us. What we really care about is that the compiler doesn't make any invalid assumptions about the stability of memory when performing optimizations. E.g., consider the following simple program:

```rust
if len < MAX_LEN {
    // do stuff
}
if len < MAX_LEN {
    // do more stuff
}
```

Assuming that `len` is never modified by the current thread, the compiler can assume that the second bounds check is spurious and elide it. If, however, `len` is actually stored in shared memory and the remote process concurrently modifies it, that optimization would be invalid. In practice, of course, the issues would likely be much more complicated and subtle.

**Comment by joshlf** [comment]
> Yes. Exactly. What you are describing is what _volatile_ can handle and what atomic cannot. Please read the above llvm example again. Repeated atomic operations _can be elided_ by the compiler if it chooses to.

The sorts of elisions that you cite (such as eliding back-to-back sequential stores) affect neither security nor correctness. They don't affect security since all we're trying to avoid is UB, and LLVM says it's not UB to use atomic operations on memory which is being concurrently modified. They don't affect correctness because, in the case of eliding back-to-back sequential stores, a) the current thread can't observe the value of the first store and, b) other threads cannot _rely_ on observing the value of the first store before it's overwritten by the second, and so optimizing to only use the second store does not violate any guarantees.

> That said, unsynchronized volatile modification is also UB (data race)

Ah, I didn't know that. I'll have to reconsider.

**Comment by jynnantonix** [comment]
I found this issue because I was also looking for a way to read a slice of structs from volatile memory.  It turns out that since `read_volatile` and `write_volatile` allow you to read and write arrays of arbitrary sizes, you can implement `{read,write}_volatile_slice` methods using them:

```rust
macro_rules! volatile_read_array {
    ($src:expr, $dst:expr, $ty:ty, $count:expr) => {
        volatile_read_array!(
            $src, $dst, $ty, $count, 1024, 512, 256, 128, 64, 32, 16, 8, 4, 2, 1
        )
    };
    ($src:expr, $dst:expr, $ty:ty, $count:expr, $default:expr, $($N:expr),* $(,)?) => {
        match $count {
            $(
                c if c >= $N => {
                    ptr::write($dst as *mut [$ty; $N], ptr::read_volatile($src as *const [$ty; $N]));
                    $N
                }
            )*
            _ => unreachable!("count is 0"),
        }
    };
}

unsafe fn volatile_read_memory<T>(mut src: *const T, mut dst: *mut T, mut count: usize) {
    while count > 0 {
        let copied = volatile_read_array!(src, dst, T, count);
        src = src.add(copied);
        dst = dst.add(copied);
        count -= copied;
    }
}

macro_rules! volatile_write_array {
    ($src:expr, $dst:expr, $ty:ty, $count:expr) => {
        volatile_write_array!(
            $src, $dst, $ty, $count, 1024, 512, 256, 128, 64, 32, 16, 8, 4,  2, 1
        )
    };
    ($src:expr, $dst:expr, $ty:ty, $count:expr, $default:expr, $($N:expr),* $(,)?) => {
        match $count {
            $(
                c if c >= $N => {
                    write_volatile($dst as *mut [$ty; $N], ptr::read($src as *const [$ty; $N]));
                    $N
                }
            )*
            _ => unreachable!("count is 0"),
        }
    };
}

unsafe fn volatile_write_memory<T>(mut src: *const T, mut dst: *mut T, mut count: usize) {
    while count > 0 {
        let copied = volatile_write_array!(src, dst, T, count);
        src = src.add(copied);
        dst = dst.add(copied);
        count -= copied;
    }
}
```

In this case, I think it would make more sense to stabilize the intrinsics instead of making people work around it like this.

**Comment by Demindiro** [comment]
> That said, unsynchronized volatile modification is also UB (data race)

Would that not make any MMIO access UB? MMIO access is usually done with volatile reads/writes without any explicit synchronization. (Notably, the GMBUS2 register in Intel HD Graphics GPUs has a bit that is intended to be used as a semaphore by software only.)
